bindings/rabbitmq/rabbitmq.go

```diff
@@ -230,66 +291,153 @@ func (r *RabbitMQ) declareQueue() (amqp.Queue, error) {
 		args[rabbitMQMaxPriorityKey] = *r.metadata.MaxPriority
 	}
 
-	return r.channel.QueueDeclare(r.metadata.QueueName, r.metadata.Durable, r.metadata.DeleteWhenUnused, r.metadata.Exclusive, false, args)
+	return channel.QueueDeclare(r.metadata.QueueName, r.metadata.Durable, r.metadata.DeleteWhenUnused, r.metadata.Exclusive, false, args)
 }
 
 func (r *RabbitMQ) Read(ctx context.Context, handler bindings.Handler) error {
 	if r.closed.Load() {
 		return errors.New("binding already closed")
 	}
 
-	msgs, err := r.channel.Consume(
-		r.queue.Name,
-		"",
-		false,
-		false,
-		false,
-		false,
-		nil,
-	)
-	if err != nil {
-		return err
-	}
-
 	readCtx, cancel := context.WithCancel(ctx)
-
 	r.wg.Add(2)
 	go func() {
-		defer r.wg.Done()
-		defer cancel()
 		select {
 		case <-r.closeCh:
+			// nop
 		case <-readCtx.Done():
+			// nop
 		}
+		r.wg.Done()
+		cancel()
 	}()
-
 	go func() {
+		// unless closed, keep trying to read and handle messages forever
 		defer r.wg.Done()
-		var err error
 		for {
-			select {
-			case <-readCtx.Done():
+			var (
+				msgs              <-chan amqp.Delivery
+				err               error
+				declaredQueueName string
+				ch                *amqp.Channel
+			)
+			r.channelMutex.RLock()
+			declaredQueueName = r.queue.Name
+			ch = r.channel
+			r.channelMutex.RUnlock()
+			if ch == nil {
+				goto Retry
+			}
+			msgs, err = ch.Consume(
+				declaredQueueName,
+				"",
+				false,
+				false,
+				false,
+				false,
+				nil,
+			)
+			if err != nil {
+				r.logger.Errorf("Error consuming messages from queue [%s]: %v", r.queue.Name, err)
+				goto Retry
+			}
```

Comment:
To prevent us from needing to use a `goto`. This also removes the sleep below which currently always prevents `Close` from returning up to `r.metadata.reconnectWait` duration.

```suggestion
			if err != nil {
				r.logger.Errorf("Error consuming messages from queue [%s]: %v", r.queue.Name, err)
				select {
				  case <-time.After(r.metadata.reconnectWait):
				  case <-readCtx.Done():
				    r.logger.Info("Input binding closed, stop fetching message")
				    return
				}
			}
```

Reply:
To avoid the goto and the duplicate select, I slightly changed the code. When there's any error or `r.handleMessage` finished, it will eventually execute the select part.